pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,972 @@
+	tokenInterface := sc.Management.Tokens("")
+	tokenList, err := tokenInterface.List(metav1.ListOptions{})
+	if err != nil {
+		logrus.Errorf("[%v] unable to fetch token objects: %v", migrateAdUserOperation, err)
+		return err
+	}
+
+	for i, workunit := range *workunits {
+		guidPrincipal := activeDirectoryPrefix + workunit.guid
+		for _, token := range tokenList.Items {
+			if guidPrincipal == token.UserPrincipal.Name {
+				workunit.guidTokens = append(workunit.guidTokens, token)
+			} else {
+				for _, duplicateLocalUser := range workunit.duplicateUsers {
+					if localPrincipalID(duplicateLocalUser) == token.UserPrincipal.Name {
+						workunit.duplicateLocalTokens = append(workunit.duplicateLocalTokens, token)
+					}
+				}
+			}
+		}
+		(*workunits)[i] = workunit
+	}
+
+	return nil
+}
+
+func collectCRTBs(workunits *[]migrateUserWorkUnit, sc *config.ScaledContext) error {
+	crtbInterface := sc.Management.ClusterRoleTemplateBindings("")
+	crtbList, err := crtbInterface.List(metav1.ListOptions{})
+	if err != nil {
+		logrus.Errorf("[%v] unable to fetch CRTB objects: %v", migrateAdUserOperation, err)
+		return err
+	}
+
+	for i, workunit := range *workunits {
+		guidPrincipal := activeDirectoryPrefix + workunit.guid
+		for _, crtb := range crtbList.Items {
+			if guidPrincipal == crtb.UserPrincipalName {
+				workunit.guidCRTBs = append(workunit.guidCRTBs, crtb)
+			} else {
+				for _, duplicateLocalUser := range workunit.duplicateUsers {
+					if localPrincipalID(duplicateLocalUser) == crtb.UserPrincipalName {
+						workunit.duplicateLocalCRTBs = append(workunit.duplicateLocalCRTBs, crtb)
+					}
+				}
+			}
+		}
+		(*workunits)[i] = workunit
+	}
+
+	return nil
+}
+
+func collectPRTBs(workunits *[]migrateUserWorkUnit, sc *config.ScaledContext) error {
+	prtbInterface := sc.Management.ProjectRoleTemplateBindings("")
+	prtbList, err := prtbInterface.List(metav1.ListOptions{})
+	if err != nil {
+		logrus.Errorf("[%v] unable to fetch PRTB objects: %v", migrateAdUserOperation, err)
+	}
+
+	for i, workunit := range *workunits {
+		guidPrincipal := activeDirectoryPrefix + workunit.guid
+		for _, prtb := range prtbList.Items {
+			if guidPrincipal == prtb.UserPrincipalName {
+				workunit.guidPRTBs = append(workunit.guidPRTBs, prtb)
+			} else {
+				for _, duplicateLocalUser := range workunit.duplicateUsers {
+					if localPrincipalID(duplicateLocalUser) == prtb.UserPrincipalName {
+						workunit.duplicateLocalPRTBs = append(workunit.duplicateLocalPRTBs, prtb)
+					}
+				}
+			}
+		}
+		(*workunits)[i] = workunit
+	}
+
+	return nil
+}
+
+func migrateCRTBs(workunit *migrateUserWorkUnit, sc *config.ScaledContext, dryRun bool) error {
+	crtbInterface := sc.Management.ClusterRoleTemplateBindings("")
+	// First convert all GUID-based CRTBs to their equivalent Distinguished Name variants
+	dnPrincipalID := activeDirectoryPrefix + workunit.distinguishedName
+	for _, oldCrtb := range workunit.guidCRTBs {
```

Comment:
A couple of things here:
- First, I would warn that migrating a CRTB involves deleting the old CRTB and making a new one with a different name. This will break people's state files (who use things like terraform), and they should know about that.
- Second, I would recommend putting the name of the previous CRTB on the new CRTB as a label. This will allow users to track what CRTB was transformed into another CRTB, and potentially even automate state reconciliation, which is a big plus.

Reply:
Mostly outside the script (in a readme). However, the dry run output should be clear that the old CRTB will be deleted and that the new CRTB will have a different name - not sure that it is clear on that today.

Reply:
> Mostly outside the script (in a readme). However, the dry run output should be clear that the old CRTB will be deleted and that the new CRTB will have a different name - not sure that it is clear on that today.

README added.  Also noted in banner when running manually. Will also be part of release notes.